### DISCUSSION TITLE
Add a parameter to router.push to force SSR for a single request instead of the default ISR set in getStaticPaths

### DISCUSSION THREAD

**Discussion by cirosantilli** [vercel/next.js#25677] [Ideas]
EDIT: @sergioengineer mentions at https://github.com/vercel/next.js/discussions/25677#discussioncomment-1841394 that this can be done with Preview mode from Next.js 12: https://nextjs.org/docs/advanced-features/preview-mode Fantastic! I wish it had been mentioned more clearly on the release blog at: https://nextjs.org/blog/next-12 Note that preview mode views don't update the cache, so the first visitor that comes to the page without cookies, including potentially the author themself, will also see the old page once. Cache invalidate/update is a very big ask by many people, related: https://stackoverflow.com/questions/68062987/next-js-isr-incremental-static-regeneration-how-to-rebuild-or-update-a-spec

The use case is as follows: I have a multi-user blog (mini Medium clone, prototype at https://github.com/cirosantilli/node-express-sequelize-nextjs-realworld-example-app/tree/63150b632f5e6feda8ba3b8029e22de25543f162 which is a port of the awesome https://github.com/gothinkster/realworld )

The website allows users to create and edit posts from `/edit/` and `/edit/[post-id]`. After the user creates or edits a post, they should be redirected to the post view `/view/[post-id]`, which gives them a chance to see if everything looks good.

I would like to use ISR by default to get faster load times and save server usage when readers access the page `/view/[post-id]` by default, thus returning from `getStaticPaths` keys `fallback: true` and `revalidate`.

The problem with this is that when the user edits an existing post and gets redirected to the view page, the backend database does get updated via my custom API, but they will initially see the old version of of the page, since it is that initial visit that triggers the cache rebuild. They would only see the updated page if they manually refresh the page after the cache is updated.

This behavior would be confusing to post authors, who would be left thinking: "did the update take place or not".

I wonder if we could add an option to `router.push`, which gets called after the user clicks the "Submit update" button which forces, just for one time, for the view page `/view/[post-id]` to do SSR, i.e. make the browser either block until reply (like `fallback: 'blocking'`, or better show the loading following by automatically updating when render is done like `fallback: true`.

Or is there already a simple way to achieve this behavior?

I have not idea about implementation details :-) I think this would require sending a custom HTTP request to the server to force `revalidate=0` for a single request, and then browser-side it should be just doing the same as `blocking` for a single request. If this gets approved I can try to dig into the code and attempt an implementation.

Related discussion I could find: https://github.com/vercel/next.js/discussions/11698#discussioncomment-351289

**If this gets done, we would also like a router option to send the updated page now request without moving to it**

The use case is as follows: you are going to leave a comment in a blog post. Upon clicking publish, you want your comment to appear without page reload.

Currently this can be implemented by making the backend server API return a JSON version of the comment, which has to look identical to the one setup by `getStaticProps`. After submiting the new comment, you want for this reply, and add it to the list of comments. You then make the commentList received by the render function part of the render state, and update that list with the return. This is what I actually ended up going for, because I would need to do an API request every time anyways to show information that is specific to the logged in user, e.g. "did the logged in user previously like this article?": https://github.com/cirosantilli/node-express-sequelize-nextjs-realworld-example-app/blob/371583cf97fb3bac16be26652e3d17f59782a166/pages/article/%5Bpid%5D.tsx#L27

However, this won't trigger the page rebuild. So if e.g. the user clicks away, and then clicks back, only the backclick will kickstart regeneration, their comment won't be there initially unless they manually reload the page.

So what we want is after the new comment POST is sent, to make the client send that "update page now request". Even now, it could still happen that the user sees the old page for a short while if it falls withing the `revalidate` period. To solve that completely we would need to actually mark a per-page one-time '`blocking'`... but at least it is much less likely to happen.

**Comment by tsaopei** [comment]
+1 on this request - I have the same issue here and I imagine the same pattern applies to anyone working on a creator economy type product, where the author of any content wants to preview their change after submitting the change.

**Comment by schoening** [comment]
This would be a welcomed addition. I would like to set up a custom webhook that triggers rebuilds when sanity data is updated (sanity has a webhook that can inform a server about the change)

Perhaps a solution that uses an URL path would be the simplest solution for users to understand?
```
next.rebuild('/my-blog/hello-world')
```

**Reply by matepaiva** [reply]
@schoening this is exactly what I need. Meanwhile, did you find a workaround to it?

**Reply by schoening** [reply]
Unfortunately not...
I currently don't have time to look where this could be added in NextJS, and switched over to elm-pages 2.0 for my current project. But hopefully voicing a need about this puts it on the roadmap of the developers 😊
elm-pages also does not have this feature and I was told that it is a complex problem to solve. I clearly don't understand it enough about the proven, because I would think that at the stage of creating a list of routes to be created that list could be filtered down to the needed one and then the build process continued. It's probably more difficult.

**Comment by iampava** [comment]
+1. Any updates on this?

**Reply by cirosantilli** [reply]
Unfortunately not, CRUD websites does not seem to be a major priority of Next.js.

**Reply by sergioengineer** [reply]
@cirosantilli, @iampava
There's an option for that called [preview mode](https://nextjs.org/docs/advanced-features/preview-mode).
They added it in Nextjs's 12th version if I'm not mistaken so you might have to migrate to the latest version.

You have to call **setPreviewData** inside a preview API. This will set a magic cookie that signals the server to ignore cache and rebuild the page by calling **getStaticProps** with **context.preview = true** essentially making **getStaticProps** behave like **getServerSideProps** for however long you specify that cookie to live, through the **maxAge** prop, or until you clear the cookies. 

I never used this feature, but I believe that you have to secure your preview API so that users/bots don't get to just force your server to rebuild the cache ad infinitum.

**Reply by cirosantilli** [reply]
Oh, fantastic!!
